_posts/2017-09-04-Secure-your-architecture-part1.md

```diff
@@ -0,0 +1,427 @@
+<dependency>
+    <groupId>org.springframework.cloud</groupId>
+    <artifactId>spring-cloud-starter-oauth2</artifactId>
+</dependency>
+{% endhighlight %}
+
+To catch the OAuth2 token we have to add the extra `@EnableResourceServer` annotation and because we are using JWT, we need an extra implementation to store and convert our JWT.
+The OAuth2 starter includes the JWT dependency of Spring Security.
+Secondly, we make a class to decode our incoming JWT.
+
+{% highlight java %}
+
+public class JwtAccessTokenConvertDecoder extends JwtAccessTokenConverter {
+
+    private JsonParser objectMapper = JsonParserFactory.create();
+
+    protected Map<String, Object> decode(String token) {
+        try {
+            Jwt jwt = JwtHelper.decode(token);
+            String content = jwt.getClaims();
+            Map<String, Object> map = objectMapper.parseMap(content);
+            if (map.containsKey(EXP) && map.get(EXP) instanceof Integer) {
+                Integer intValue = (Integer) map.get(EXP);
+                map.put(EXP, Long.valueOf(intValue));
+            }
+            return map;
+        }
+        catch (Exception e) {
+            throw new InvalidTokenException("Cannot convert access token to JSON", e);
+        }
+    }
+}
+
+{% endhighlight %}
+
+And add these as beans to get it working. 
+
+{% highlight java %}
+@Configuration
+@EnableResourceServer
+public class ResourceServiceConfiguration extends ResourceServerConfigurerAdapter {
+
+    @Bean
+    public JwtAccessTokenConvertDecoder accessTokenConverter() {
+        return new JwtAccessTokenConvertDecoder();
+    }
+
+    @Bean
+    public TokenStore tokenStore() {
+        return new JwtTokenStore(accessTokenConverter());
+    }
+
+    @Bean
+    @Primary
+    public DefaultTokenServices tokenServices() {
+        DefaultTokenServices defaultTokenServices = new DefaultTokenServices();
+        defaultTokenServices.setTokenStore(tokenStore());
+        return defaultTokenServices;
+  }
+}
+
+{% endhighlight %}
+
+### Common mistakes with keys
+So the UAA signs the token with a private key and verifies it with a public key. 
+The problem that occurs is that every microservice would need the public key to connect with the UAA server for verification on every request.
+As you can see, we don't want every microservice to be dependant of the UAA server to be available at startup, in testing or CI. 
+The solution to this is to disable exposure of your microservices to the outer network and handle only incoming traffic via the reverse proxy/load balancer (Zuul).
+Zuul will verify the token as a trustworthy client of the UAA server and will propagate the token to the downstreamed services.
+The microservice will handle the token as trustworthy since the token is verified.
+
+
+### Securing API endpoints
+At last we're going to secure our resources.
+Spring Security gives us a variety of tools to secure at class/method level.
+Let me point you out the most used one.  <br />
```

Comment:
Sowieso de "you" droppen, maar de hele constructie klinkt een beetje vreemd.
Misschien eerder iets in de aard van "The one that's used most often enables method security, which you enable by adding `@EnableGlobalMethodSecurity(prePostEnabled = true)` to your configuration"